src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -171,18 +246,19 @@ MemoryScope acquire() {
         }
 
         @Override
-        MemoryScope dup() { // always called in owner thread
-            return closeOrDup(false);
+        MemoryScope dup(Thread owner) {
+            Objects.requireNonNull(owner, "owner");
+            return closeOrDup(owner);
         }
 
         @Override
-        void close() { // always called in owner thread
-            closeOrDup(true);
+        void close() {
+            closeOrDup(null);
         }
 
-        private MemoryScope closeOrDup(boolean close) {
+        private MemoryScope closeOrDup(Thread newOwner) {
```

Comment:
I think I don't like this way of sharing the code. Since we're in cleanup mode, I think the code would be more readable by having two separate methods `close` and `dup` which both uses the same underlying logic for flipping the liveness bit (using the lock).